src/main/scala/edu/ie3/simona/model/participant2/HpModel.scala

```diff
@@ -0,0 +1,493 @@
+
+    val (newHpActivePower, _, qDotIntoGrid) = {
+      if (turnOn)
+        (pRated, pThermal, pThermal)
+      else if (
+        currentStorageEnergy > zeroKWh && state.thermalDemands.houseDemand.hasRequiredDemand
+      ) {
+        // If the house has req. demand and storage isn't empty, we can heat the house from storage.
+        (zeroKW, zeroKW, storagePThermal)
+      } else if (
+        currentStorageEnergy > zeroKWh && state.thermalDemands.houseDemand.hasPossibleDemand && lastHouseQDot > zeroKW
+      )
+        // Edge case when em controlled: If the house was heated last state by Hp and setPower is below turnOn condition now,
+        // but house didn't reach target or boundary temperature yet. House can be heated from storage, if this one is not empty.
+        (zeroKW, zeroKW, storagePThermal)
+      else (zeroKW, zeroKW, zeroKW)
+    }
+
+    (newHpActivePower, qDotIntoGrid)
+  }
+
+  override def createResults(
+      state: HpState,
+      lastOperatingPoint: Option[HpOperatingPoint],
+      currentOperatingPoint: HpOperatingPoint,
+      complexPower: ComplexPower,
+      dateTime: ZonedDateTime,
+  ): Iterable[ResultEntity] = {
+    Iterable(
+      new HpResult(
+        dateTime,
+        uuid,
+        complexPower.p.toMegawatts.asMegaWatt,
+        complexPower.q.toMegavars.asMegaVar,
+        currentOperatingPoint.thermalOps
+          .map(_.qDotHp)
+          .getOrElse(
+            throw new RuntimeException(
+              s"currentOperatingPoint $currentOperatingPoint does not contain heat value but one is expected."
+            )
+          )
+          .toMegawatts
+          .asMegaWatt,
+      )
+    ) ++ thermalGrid.results(
+      state,
+      lastOperatingPoint,
+      currentOperatingPoint,
+      dateTime,
+    )
+  }
+
+  override def createPrimaryDataResult(
+      data: PrimaryDataWithComplexPower[_],
+      dateTime: ZonedDateTime,
+  ): SystemParticipantResult = {
+    data match {
+      case result: ComplexPowerAndHeat =>
+        new HpResult(
+          dateTime,
+          uuid,
+          result.p.toMegawatts.asMegaWatt,
+          result.q.toMegavars.asMegaVar,
+          result.qDot.toMegawatts.asMegaWatt,
+        )
+    }
+  }
+
+  override def getRequiredSecondaryServices: Iterable[ServiceType] =
+    Iterable(ServiceType.WeatherService)
+
+  /** Calculate the active power behaviour of the model.
+    *
+    * @param state
+    *   The current state including weather data.
+    * @return
+    *   The active power.
+    */
+
+  override def determineOperatingPoint(
+      state: HpState
+  ): (HpOperatingPoint, Option[Long]) = {
+
+    val (newActivePowerHp, qDotIntoGrid) = nextOperatingPoint(state, None)
+
+    /* Push thermal energy to the thermal grid and get its updated state in return */
+    val (updateState, maybeThreshold) = thermalGrid.updateState(
+      state.tick,
+      state,
+      newActivePowerHp > zeroKW,
+      qDotIntoGrid,
+      state.thermalDemands,
+    )
```

Comment:
Hmm, this also doesn't seem like I intended this. When we're determining the operating point, the state of the grid should already be determined: The parameter `state` is already the current state.

Reply:
I guess, we could adapt this. We only need to get back the information how the qDotHp is used from the thermalGrid (for the house, for the storage, so split?) and the next threshold.

But like this, it seems to me more compact. Other thing, is that I now would like to find a better naming for updateState() since there is now also updatedThermalGridState(), I also tried to combine both, but this need a bit more refactoring, I guess.

Reply:
Writing this i got some idea :)